### ISSUE TITLE
Questions regarding pgBouncer

### ISSUE THREAD

**Issue by v3nturetheworld** [CrunchyData/postgres-operator#1411]
**Which example are you working with?**

I'm not working with an example but I'm trying to setup a basic, single replica cluster with pgBouncer for connection pooling to avoid database disconnects on fail overs. The cluster was created with the following command:
`pgo create cluster -n crunchy-test hacluster --pod-anti-affinity="preferred" --replica-count=1 --pgbouncer`

**What is the current behavior?**
All of the services and pods associated with the cluster correctly run, however the pgBouncer doesn't seem to connect to either the primary or secondary database. 

**What is the expected behavior?**
Unless there is additional required configuration that I am missing (I'm new to pgBouncer and working with PostgreSQL), I believe I should be able to use psql to connect through the pgBouncer Service to the databases such as `userdb` using `psql -U pgbouncer -p 5432 -h hacluster-pgbouncer userdb`

**Other information** (e.g. detailed explanation, related issues,  etc)
From within the pgBouncer pod I am able to use psql to connect directly to the hacluster service using: `psql -U postgres -p 5432 -h hacluster userdb`. I am also able to connect to the 'virtual database' pgBouncer has. I was testing a few queries to see if I was missing something configuration wise, but if I should try something else here please let me know:

![image](https://user-images.githubusercontent.com/4187476/79281329-216a2080-7e78-11ea-9e43-6c2c012e205b.png)

![image](https://user-images.githubusercontent.com/4187476/79281435-52e2ec00-7e78-11ea-9e5d-a28222c3e663.png)

![image](https://user-images.githubusercontent.com/4187476/79281450-5aa29080-7e78-11ea-9049-9294fc81544b.png)

Trying to test creating a connection to the userdb within the pgbouncer container resulted in this:
![image](https://user-images.githubusercontent.com/4187476/79281821-33988e80-7e79-11ea-92f7-caa212854afc.png)

Next, I tested from the hacluster-replica container:
![image](https://user-images.githubusercontent.com/4187476/79281889-5a56c500-7e79-11ea-8482-2cde5cf3127f.png)

I may be missing something obvious here, or maybe its my lack of understanding on how pgBouncer works, but from the documentation and all that I can find online, adding the `--pgbouncer` flag seems like it should automatically do all the configuration needed. For an application I should be able to point it to the pgBouncer Service and I guess either provide the pgbouncer credentials or by using `pgo create user` use those credentials for the application. One additional change I made was adding the replica configuration option to `pgbouncer.ini` with: `{{.PG_REPLICA_SERVICE_NAME}} = host={{.PG_REPLICA_SERVICE_NAME}} port={{.PG_PORT}} auth_user={{.PG_USERNAME}} dbname={{.PG_DATABASE}}`

Additionally, I noticed in the Environment variables for pgBouncer deployment that PG_PRIMARY_SERVICE_NAME and PG_REPLICA_SERVICE_NAME were both empty. Should these fields be populated?
![image](https://user-images.githubusercontent.com/4187476/79282390-ace4b100-7e7a-11ea-804e-94cc0441ac42.png)

**Please tell us about your environment:**

* Operating System:
Docker images are CentOS7, cluster OS is RHEL7
* Where is this running ( Local , Cloud Provider): Local 3 node OpenShift3.11 Cluster.
* Storage being used (NFS, Hostpath, Gluster, etc): NFS
* Container Image Tag: centos7-4.2.2 and centos7-12.2-4.2.2
* PostgreSQL Version: 12.2
* Platform (Docker, Kubernetes, OpenShift): OpenShift
* Platform Version: 3.11

If possible please run the following on the kubernetes or OpenShift (oc) commands and provide the result:
 kubectl describe yourPodName
 kubectl describe pvc
 kubectl get nodes
 kubectl log yourPodName

![image](https://user-images.githubusercontent.com/4187476/79282459-e0274000-7e7a-11ea-90b7-0b337e19e8d4.png)

![image](https://user-images.githubusercontent.com/4187476/79282477-eb7a6b80-7e7a-11ea-9b1b-7604f2825449.png)

**Comment by jkatz** [comment]
The short answer is that in the [upcoming release](https://github.com/CrunchyData/postgres-operator/releases/tag/v4.3.0-beta.1) the pgBouncer integration received significant improvements, including some of the foundational ones found in c31fae36

In 4.2, For connecting to "any" database within your PostgreSQL cluster, you want to use the * setting, as such:

[https://github.com/CrunchyData/postgres-operator/blob/master/conf/postgres-operator/pgbouncer.ini#L1-L2](https://github.com/CrunchyData/postgres-operator/blob/master/conf/postgres-operator/pgbouncer.ini#L1-L2)

You can edit this file in the `pgo-config` ConfigMap and then redeploy your pgBouncer Deployment.

If you installed with the Ansible installer, it appears the two pgbouncer.ini files have deviated. I am going to bring them into alignment.

**Comment by v3nturetheworld** [comment]
@jkatz Thanks for getting back to me. I look forward to the 4.3 release, it looks like it will have a lot of great new features and improvements, I particularly look forward to the installation container.

I made the change to config map you suggested but I think I may still be missing something. If I understand the pgBouncer correctly, an application would use the pgbouncer username and password, along with the pgbouncer service hostname to connect to the databases in the cluster, correct? Is any other additional configuration required, or should the `--pgbouncer` flag be enough in `pgo create cluster` command? 

So I'm trying to connect to the userdb database in the master database via the pgbouncer (using psql in pgbouncer pod), and I get the following results:
![image](https://user-images.githubusercontent.com/4187476/79376947-8ed18c00-7f20-11ea-831f-4041ce2005e4.png)

Do I have to expose ports anywhere using oc commands or edit any additional files?

As a side note, I believe this may be a bug but I discovered if you delete the pgbouncer in a cluster then create a new pgbouncer, the new pgbouncer never gets deployed.

**Comment by jkatz** [comment]
I appreciate the very detailed feedback and logs, this certainly helps a lot when trying to troubleshoot.

That being said, I will give you an answer that may not be totally satisfying, but will make your pgBouncer experience much more enjoyable. Wait for 4.3 :smile: A lot of what you mentioned above was fixed in this upcoming release as the pgBouncer support was completely re-engineered.

To try to address a few of the above, regarding:

```ini
* = host={{.PG_PRIMARY_SERVICE_NAME}} port={{.PG_PORT}} auth_user=pgbouncer
```

This basically says "connect to any database via the primary service over the supplied port using the `pgbouncer` user to deliver the authentication credentials for my other users" (bear with me on this one).

What happens is that there is a special `SECURITY DEFINER` function that queries the `pg_catalog.pg_authid` table to grab the hashed credentials for users that can log into the PostgreSQL database. This allows both pgBouncer to verify the person who is logging in can actually log in.

In 4.2, I would not be surprised if the pgBouncer user was not added to the database -- this is all fixed in 4.3, as I personally spent well over a week tightening up our pgBouncer integration.

That all said, you should be able to create user with the `pgo create user` command in general. They may not be immediately available for logging in via pgBouncer (or you may need to bounce the Pod, no pun intended). I also spent some time on `pgo create user` in 4.3 as well which coincides with the pgBouncer improvements too.

Anyway,in a few weeks, this will make much more sense :smile:

**Comment by v3nturetheworld** [comment]
Awesome, thanks again for the reply. I figured something like that was happening with the user accounts in pgBouncer and the databases after reading a bit more on what pgBouncer does and how it works. I think for now I'll just go without the bouncer and use the primary db service name and live with the 30-40s time for failover to occur until 4.3 gets released. I'll close this issue now and look forward to the release of 4.3... so far Crunchy has been pretty awesome though 👍.
